### PR TITLE
Add ABIFM to IceNucleation

### PR DESCRIPTION
**Pull request by amylu00**
## Purpose 

Adding immersion freezing nucleation rate for sulphuric acid droplets.

## To-do

- fix performance test for H2SO4 sat vapor pressure

## Content

- added ABIFM derived J to IceNucleation.jl
- added ABIFM derived water activity to IceNucleation.jl
- added m and c values to calculate J for 3 aerosol types in IceNucleation.jl
- added documentation for ABIFM in IceNucleation.md
- added comparison to Knopf & Alpert 2013 fig 4 in IceNucleation.md
- added performance tests

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: docs/src/IceNucleationBox.md [hunk 1]
```diff
@@ -8,11 +8,6 @@ It is based on [KorolevMazin2003](@cite), as well as the cirrus box model
 The model solves for saturation in a 0-dimensional
   adiabatic parcel raising with constant velocity.
 
-!!! note
-    For now, the equations are written assuming
```
Comment (trontrytel) [inline docs/src/IceNucleationBox.md:12]: Could we also separate this one from the ABIFM PR?

Comment (trontrytel) [reply docs/src/IceNucleationBox.md:12]: Also, would be great to add a parcel example where we just have liquid water grow via condensation using the non-equilibrium microphysics. And show that we can reproduce the typical supersaturation(height) behavior

File: parcel/ABIFMparcel.jl [hunk 2]
```diff
@@ -0,0 +1,230 @@
+import OrdinaryDiffEq as ODE
+import CairoMakie as MK
+
```
Comment (trontrytel) [inline parcel/ABIFMparcel.jl:3]: I would prefer to also add this file in a separate PR with parcel model updates

File: parcel/parcel.jl [hunk 3]
```diff
@@ -45,6 +45,7 @@ function cirrus_box(dY, Y, p, t)
     R_a = TD.gas_constant_air(thermo_params, q)
     cp_a = TD.cp_m(thermo_params, q)
     L_subl = TD.latent_heat_sublim(thermo_params, T)
+    L_fus = TD.latent_heat_fusion(thermo_params, T)
     ρ = TD.air_density(thermo_params, ts)
```
Comment (trontrytel) [inline parcel/parcel.jl:49]: Same here

File: src/IceNucleation.jl [hunk 4]
```diff
@@ -54,4 +55,39 @@ function dust_activated_number_fraction(
     end
 end
 
+"""
+   ABIFM_J(prs, x, T)
+
+ - `prs` - set with model parameters
+ - `x` - wt% sulphuric acid / 100
```
Comment (trontrytel) [inline src/IceNucleation.jl:62]: Could you add some more explanation on what this variable is and what are the units?

File: src/IceNucleation.jl [hunk 5]
```diff
@@ -54,4 +55,39 @@ function dust_activated_number_fraction(
...
+
+    thermo_params = CMP.thermodynamics_params(prs)
+
+    m = 54.48075    # for illite particles
+    c = -10.66873
+
+    w_h = 1.4408 * x
+    p_sol = exp( 23.306 - 5.3465*x + 12*x*w_h - 8.19*x*w_h^2 + ( -5814 + 928.9*x - 1876.7*x*w_h )/T ) * 100 # * 100 converts mbar --> Pa
+    p_sat = TD.saturation_vapor_pressure(thermo_params, T, TD.Liquid())
+    p_ice = exp( 9.550426 - 5723.265/T + 3.53068*log(T) - 0.00728332*T ) # check with TD.saturation_vapor_pressure(thermo_params, T, TD.Ice())
```
Comment (trontrytel) [inline src/IceNucleation.jl:82]: We have saturation over ice defined in TD.jl

File: src/IceNucleation.jl [hunk 6]
```diff
@@ -54,4 +55,39 @@ function dust_activated_number_fraction(
...
+"""
+function ABIFM_J(
+    prs::APS,
+    x::FT,
+    T::FT
+) where {FT <: Real}
+
+    thermo_params = CMP.thermodynamics_params(prs)
+
+    m = 54.48075    # for illite particles
```
Comment (trontrytel) [inline src/IceNucleation.jl:76]: We will want to make those parameters part of CLIMAParameters at some point. Maybe for now we can just define them at the top of the file. So that they are easier to move later on

File: src/IceNucleation.jl [hunk 7]
```diff
@@ -54,4 +55,39 @@ function dust_activated_number_fraction(
...
+    c = -10.66873
+
+    w_h = 1.4408 * x
+    p_sol = exp( 23.306 - 5.3465*x + 12*x*w_h - 8.19*x*w_h^2 + ( -5814 + 928.9*x - 1876.7*x*w_h )/T ) * 100 # * 100 converts mbar --> Pa
```
Comment (trontrytel) [inline src/IceNucleation.jl:80]: Maybe we could define it as a separate function?

File: parcel/ABIFMparcel.jl (hunk 2 above)
Comment (amylu00) [reply parcel/ABIFMparcel.jl:3]: For now, I'm using this file as a test to try to reproduce the figure in Knopf & Alpert. Once I confirm that the ABIFM J function works correctly, I was going to delete it. Would it be better if I wait until the parcel model updates on the parcel branch are done before trying to do this?

File: src/IceNucleation.jl (hunk 7 above)
Comment (amylu00) [reply src/IceNucleation.jl:80]: I could do that on the same file, unless you want it in Thermodynamics so it's more consistent with p_sat and p_ice?

File: parcel/ABIFMparcel.jl (hunk 2 above)
Comment (trontrytel) [reply parcel/ABIFMparcel.jl:3]: Which figure are you thinking of? I thought there was one that we wanted to reproduce that did not require the parcel model. 

But if it makes it easier in testing, then lets keep it for now

File: src/IceNucleation.jl (hunk 7 above)
Comment (trontrytel) [reply src/IceNucleation.jl]: Lets keep it in CloudMicrophysics for now. But maybe in the end it will be better to move it to TD.jl

File: docs/src/IceNucleation.md [hunk 8]
```diff
@@ -2,9 +2,9 @@
...
   the parameterization of activation of dust aerosol particles into ice crystals
-  via deposition of water vapor.
-This is a heterogeneous ice nucleation process.
-The parameterization is an implementation of
+  via deposition of water vapor and water activity based immersion freezing model for sulphuric acid containing droplets.
+These are heterogeneous ice nucleation processes.
+The parameterization for deposition on dust particles is an implementation of
   the empirical formulae from [Mohler2006](@cite)
   and is valid for two types of dust particles:
   Arizona Test Dust and desert dust from Sahara.
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:10]:
```suggestion
  Arizona Test Dust and desert dust from Sahara.
  The parameterization for immersion freezing is an implementation of [KnopfAlpert2013](@cite) 
  and is valid for droplets containing sulphuric acid.
```

File: docs/src/IceNucleation.md [hunk 9]
```diff
@@ -2,9 +2,9 @@
 
 The `IceNucleation.jl` module includes
   the parameterization of activation of dust aerosol particles into ice crystals
-  via deposition of water vapor.
-This is a heterogeneous ice nucleation process.
-The parameterization is an implementation of
+  via deposition of water vapor and water activity based immersion freezing model for sulphuric acid containing droplets.
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:5]:
```suggestion
  via deposition of water vapor and water activity based parameterization of immersion freezing.
```

File: docs/src/IceNucleation.md [hunk 10]
```diff
@@ -42,3 +42,103 @@ Both parameters are dependent on aerosol properties and temperature.
     For a ``f_i`` values above 0.08 or ``S_i`` between 1.35 and 1.5,
     freezing occurs in a different ice nucleation mode
     (either a second deposition or other condensation type mode).
+
+## ABIFM for Sulphuric Acid Containing Droplets
+Water Activity-Based Immersion Freezing Model (ABFIM) is a method of parameterizing immersion freezing inspired by the time-dependent classical nucleation theory. The nucleation rate coefficient, ``J``, can be determined by the water activity, ``a_w``. This parameterization follows [KnopfAlpert2013](@cite), [Koop2002](@cite), [MurphyKoop2005](@cite), and [Luo1995](@cite). In this model, aerosols are assumed to contain an insoluble and soluble material. When immersed in water, the soluble material diffuses into the liquid water to create a sulphuric acid solution.
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:47]: Could you add a reference for where to read to get a good CNT overview?

Comment (trontrytel) [reply docs/src/IceNucleation.md:47]: Could you also define what ``J`` is?

File: docs/src/IceNucleation.md [hunk 11]
```diff
@@ -42,3 +42,103 @@ Both parameters are dependent on aerosol properties and temperature.
...
+
+Using empirical coefficients, ``m`` and ``c``, from [KnopfAlpert2013](@cite), the heterogeneous nucleation rate coefficient in units of ``cm^{-2}s^{-1}`` can be determined by the linear equation
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:49]: Maybe we can use the same strategy as Nat was with aerosol nucleation. I.e. write the rates in the docs as they are in the original papers. But then mention in the docs that the actual source code is in base SI units (so meters and not cm)?

File: docs/src/IceNucleation.md [hunk 12]
```diff
@@ -42,3 +42,103 @@ Both parameters are dependent on aerosol properties and temperature.
...
+  a_w = \frac{p_{sol}}{p_{sat}}
+\end{equation}
+```
+```math
+\begin{equation}
+  a_{w,ice} = \frac{p_{i,sat}}{p_{sat}}
+\end{equation}
+```
+
+where ``p`` is saturated vapor pressure of water above solution, pure liquid water, and ice for subscripts ``sol``, ``sat``, and ``i,sat`` respectively. ``p_{sol}`` is determined in mbar using a parameterization for supercooled, binary ``H_2SO_4/H_2O`` solution from [Luo1995](@cite) which is valid for 185K < T < 235K:
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:68]: `where ``p`` is saturated vapor pressure of water above solution, pure liquid water, and ice for subscripts ``sol``, ``sat``, and ``i,sat`` respectively.` - This one os a little confusing. Could you rephrase it?

File: docs/src/IceNucleation.md [hunk 13]
```diff
@@ -42,3 +42,103 @@ Both parameters are dependent on aerosol properties and temperature.
...
+```math
+\begin{equation}
+  ln(p_{sol}) = 23.306 - 5.3465x + 12xw_h - 8.19xw_h^2 + \frac{1}{T}(-5814 + 928.9x - 1876.7xw_h)
+\end{equation}
+```
+where ``x = \frac{wt%(H_2SO_4)}{100}``, ``w_h = 1.4408x``, and temperature is in Kelvins.
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:74]: I think this ```x = \frac{wt%(H_2SO_4)}{100}` displays wrongly in the final build of the documentation. Also, could you define what it is?

File: docs/src/IceNucleation.md [hunk 14]
```diff
@@ -42,3 +42,103 @@ Both parameters are dependent on aerosol properties and temperature.
...
+
+``p_{i,sat}`` in Pa follows [MurphyKoop2005](@cite) for temperatures in Kelvins above 110K:
+```math
+\begin{equation}
+  p_{i,sat} = exp[9.550426 - \frac{1}{T}5723.265 + 3.53068 ln(T) - 0.00728332T]
+\end{equation}
+```
+
+Once ``J_{het}`` is calculated, it can be used to determine the ice production rate, ``P_{ice}``, per minute via immersion freezing.
```
Comment (trontrytel) [inline docs/src/IceNucleation.md:83]: We want to stick to base SI units. So seconds not minutes

File: src/IceNucleation.jl [hunk 15]
```diff
@@ -54,4 +67,72 @@ function dust_activated_number_fraction(
     end
 end
```
Comment (trontrytel) [inline src/IceNucleation.jl:69]: In the documentation you mention temperature ranges for which the parameterization is valid. Could we add a similar warning and return zero rate for temperatures outside of that range?

Kind of similar to what we are doing here: https://github.com/CliMA/CloudMicrophysics.jl/blob/3a1b0ab630ffdefb9a83245af29931ff8ae776a6/src/IceNucleation.jl#L46-L54

**Review by trontrytel — APPROVED** [review]
This is great! Thank you so much for taking the time and adding the tests and docs. 

I left some minor comments about the documentation. Could you make sure to squash and rebase before merging?

**Comment by amylu00** [comment]
bors try

**Comment by amylu00** [comment]
bors try

**Comment by amylu00** [comment]
bors try

**Comment by amylu00** [comment]
bors try

**Comment by amylu00** [comment]
bors try
